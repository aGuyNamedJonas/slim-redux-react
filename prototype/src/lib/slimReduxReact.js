import { createSelectorCreator } from 'reselect'

/*
  Prototyping the custom selector to notify us about updates
*/

function defaultEqualityCheck(a, b) {
  return a === b
}

export function areArgumentsShallowlyEqual(equalityCheck, prev, next) {
  if (prev === null || next === null || prev.length !== next.length) {
    return false
  }

  // Do this in a for loop (and not a `forEach` or an `every`) so we can determine equality as fast as possible.
  const length = prev.length
  for (let i = 0; i < length; i++) {
    if (!equalityCheck(prev[i], next[i])) {
      return false
    }
  }

  return true
}

function defaultMemoize(func, equalityCheck = defaultEqualityCheck, hasChanged) {
  let lastArgs = null
  let lastResult = null
  // we reference arguments instead of spreading them for performance reasons
  return function () {
    let changed = false

    if (!areArgumentsShallowlyEqual(equalityCheck, lastArgs, arguments)) {
      // apply arguments instead of spreading for performance.
      lastResult = func.apply(null, arguments)
      changed = true
    }

    lastArgs = arguments

    hasChanged(changed)
    return lastResult
  }
}

export const createNotifyingSelector = (hasChanged) => createSelectorCreator(
  defaultMemoize,
  undefined,
  hasChanged,
)
