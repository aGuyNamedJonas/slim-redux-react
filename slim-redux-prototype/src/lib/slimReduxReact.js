import React from 'react';
import { getNotifyingSelectorCreator, areArgumentsShallowlyEqual } from './selector';

function slimReduxReact(params) {
  const WrappedComponent = params.component;
  const changeTriggers   = params.changeTriggers;
  const subscriptions    = params.subscriptions;

  console.log(`Subscription: ${JSON.stringify(subscriptions, null, 2)}`)

  const displayName = WrappedComponent.displayName || WrappedComponent.name || '';

  // Returns the appropriate part of the state for a string like "state.todo.active"
  const _getStateFromSubscriptionString = (subscriptionString, state) => {
    const subStringParts = subscriptionString.split('.');
    var currentPath = 'state';
    var stateFromString = state;

    for(var i=1; i < subStringParts.length; i++){
      const nextPath = subStringParts[i];
      currentPath += `.${nextPath}`;

      if(!(nextPath in stateFromString))
        console.error(`*** Error in slimReduxReact container ${displayName}:\nCannot find path "${currentPath} in state.\nState: ${JSON.stringify(state, null, 2)}"`);

      stateFromString = stateFromString[nextPath];
    }

    return stateFromString;
  }

  // Retrieves the data for the subscriptions, first part of the subscription selector
  const getSubscriptions = (state) => {
    var storeSubscriptions = {};
    Object.keys(subscriptions).map(subscription => storeSubscriptions[subscription] = _getStateFromSubscriptionString(subscriptions[subscription], state))

    console.log(`Store subscriptions: ${JSON.stringify(storeSubscriptions, null, 2)}`)
    return storeSubscriptions;
  }

  // Create subscrption selector
  const createNotifyingSelector = getNotifyingSelectorCreator();
  const checkSubscriptionSelector = createNotifyingSelector(
    getSubscriptions,
    subscriptionData => subscriptionData,
  );

  // Create change creators    TODO: change creators could also be functions to mock stuff!!
  const registerChangeTriggers = (changeTriggers, change) => {
    var registeredChangeTriggers = {}
    Object.keys(changeTriggers).map(changeTrigger => registeredChangeTriggers[changeTrigger] = change(changeTriggers[changeTrigger]))

    return registeredChangeTriggers;
  }

  class SlimReduxConnector extends React.Component {
    constructor(props, context){
      super(props);

      const initialSubscriptionState = checkSubscriptionSelector(context.store.getState())
      this.state = { ...initialSubscriptionState.data }

      this.registeredChangeTriggers = registerChangeTriggers(changeTriggers, context.store.change)
    }

    componentDidMount() {
      // SUBSCRIBE THIS MOTHERFUCKER TO SUBSCRIPTION CHANGES!
      this.context.store.subscribe(() => {
        const subscriptionState = checkSubscriptionSelector(this.context.store.getState())

        if(subscriptionState.hasChanged)
          console.log(`***UPDATED subscriptionState: ${JSON.stringify(subscriptionState, null, 2)}`)
          this.setState({
            ...this.state,
            ...subscriptionState.data,
        })
      })
    }

    render() {
      return <WrappedComponent {...this.props} {...this.registeredChangeTriggers} {...this.state}/>
    }
  }

  SlimReduxConnector.displayName = `SlimReduxConnector${displayName}`;

  SlimReduxConnector.contextTypes = {
    store: React.PropTypes.object
  }

  return SlimReduxConnector;
}

export default slimReduxReact;
