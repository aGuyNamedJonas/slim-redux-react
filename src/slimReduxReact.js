import React from 'react';
import { getNotifyingSelectorCreator, areArgumentsShallowlyEqual } from './selector';

function slimReduxReact(params) {
  const WrappedComponent = params.component;
  const changeTriggers   = params.changeTriggers || {};
  const subscriptions    = params.subscriptions || {};

  const displayName = WrappedComponent.displayName || WrappedComponent.name || '';

  // Returns the appropriate part of the state for a string like "state.todo.active"
  const _getStateFromSubscriptionString = (subscriptionString, state) => {
    const subStringParts = subscriptionString.split('.');
    var currentPath = 'state';
    var stateFromString = state;

    for(var i=1; i < subStringParts.length; i++){
      const nextPath = subStringParts[i];
      currentPath += `.${nextPath}`;

      if(!(nextPath in stateFromString))
        console.error(`*** Error in slimReduxReact container ${displayName}:\nCannot find path "${currentPath} in state.\nState: ${JSON.stringify(state, null, 2)}"`);

      stateFromString = stateFromString[nextPath];
    }

    return stateFromString;
  }

  // Retrieves the data for the subscriptions, first part of the subscription selector
  const getSubscriptions = (state) => {
    var storeSubscriptions = {};
    Object.keys(subscriptions).map(subscription => storeSubscriptions[subscription] = _getStateFromSubscriptionString(subscriptions[subscription], state))
    return storeSubscriptions;
  }

  // Create subscrption selector
  const createNotifyingSelector = getNotifyingSelectorCreator();
  const checkSubscriptionSelector = createNotifyingSelector(
    getSubscriptions,
    subscriptionData => subscriptionData,
  );

  // Create change creators    TODO: change creators could also be functions to mock stuff!!
  const registerChangeTriggers = (changeTriggers, change) => {
    var registeredChangeTriggers = {}
    Object.keys(changeTriggers).map(changeTrigger => registeredChangeTriggers[changeTrigger] = change(changeTriggers[changeTrigger]))

    return registeredChangeTriggers;
  }

  class SlimReduxConnector extends React.Component {
    constructor(props, context){
      super(props);

      if(!context.store)
        console.error(`*** Error in SlimReduxConnector component: No store found in context. Did you forget to wrap your code in the <Provider> component?`);

      const initialSubscriptionState = checkSubscriptionSelector(context.store.getState())
      this.state = { ...initialSubscriptionState.data }

      this.registeredChangeTriggers = registerChangeTriggers(changeTriggers, context.store.createChangeTrigger)
    }

    componentDidMount() {
      // SUBSCRIBE THIS MOTHERFUCKER TO SUBSCRIPTION CHANGES!
      this.context.store.subscribe(() => {
        const subscriptionState = checkSubscriptionSelector(this.context.store.getState())

        if(subscriptionState.hasChanged)
          this.setState({
            ...this.state,
            ...subscriptionState.data,
        })
      })
    }

    render() {
      return <WrappedComponent {...this.props} {...this.registeredChangeTriggers} {...this.state}/>
    }
  }

  SlimReduxConnector.displayName = `SlimReduxConnector${displayName}`;

  SlimReduxConnector.contextTypes = {
    store: React.PropTypes.object
  }

  return SlimReduxConnector;
}

export default slimReduxReact;
