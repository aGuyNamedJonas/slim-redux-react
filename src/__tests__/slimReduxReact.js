import React from 'react';
import renderer from 'react-test-renderer';
import { createSlimReduxStore } from 'slim-redux';
import slimReduxReact from '../slimReduxReact';
import { Provider } from '../index';
import { mount } from 'enzyme';
import sinon from 'sinon';

const INITIAL_COUNTER_STATE     = 0,
      INCREMENTED_COUNTER_STATE = 1,
      INCREMENT                 = 'INCREMENT';

class Counter extends React.Component {
  constructor(props){
    super(props);
  }

  render() {
    return (
      <div>
        <h1>{this.props.counter}</h1>
        <button onClick={e => this.props.increaseCounter()}>INCREASE COUNTER</button>
      </div>
    );
  }
}

const increaseCounterChangeTrigger = {
  actionType: 'INCREASE',
  reducer: state => ({ ...state, counter: state.counter + 1 }),
};

test('Provider is exported by slim-redux-react module', () => {
  expect(Provider).toBeDefined();
})

test('slimReduxReact() renders correctly', () => {
  const store = createSlimReduxStore({counter: INITIAL_COUNTER_STATE});
  const CounterContainer = slimReduxReact({
    component: Counter,
    subscriptions: { counter: 'state.counter' },
    changeTriggers: { increaseCounter: increaseCounterChangeTrigger }
  });

  const App = (
    <div>
      <Provider store={store}>
        <CounterContainer/>
      </Provider>
    </div>
  );

  const component = renderer.create(App);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot('full');
});

test('HOC renders even when no subscriptions are provided', () => {
  const store = createSlimReduxStore({counter: INITIAL_COUNTER_STATE});
  const CounterContainer = slimReduxReact({
    component: Counter,
    changeTriggers: { increaseCounter: increaseCounterChangeTrigger }
  });

  const App = (
    <div>
      <Provider store={store}>
        <CounterContainer/>
      </Provider>
    </div>
  );

  const component = renderer.create(App);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot('no-subscriptions');
});

test('HOC renders even when no change triggers are provided', () => {
  const store = createSlimReduxStore({counter: INITIAL_COUNTER_STATE});
  const CounterContainer = slimReduxReact({
    component: Counter,
    subscriptions: { counter: 'state.counter' },
  });

  const App = (
    <div>
      <Provider store={store}>
        <CounterContainer/>
      </Provider>
    </div>
  );

  const component = renderer.create(App);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot('no-change-triggers');
});

test('Wrapped component correctly receives subscribed-to data', () => {
  const store = createSlimReduxStore({counter: INITIAL_COUNTER_STATE});
  const increaseCounter = store.createChangeTrigger(increaseCounterChangeTrigger);
  const CounterContainer = slimReduxReact({
    component: Counter,
    subscriptions: { counter: 'state.counter' },
    changeTriggers: { increaseCounter: increaseCounterChangeTrigger }
  });

  const wrapper = mount(
    <div>
      <Provider store={store}>
        <CounterContainer/>
      </Provider>
    </div>
  );

  // Check what the initial counter state is inside the counter component
  const initialSubscriptionValue = wrapper.find(Counter).prop('counter');
  // Increase counter
  increaseCounter();
  // check subscribed to counter prop inside counter component
  const increaseSubscriptionValue = wrapper.find(Counter).prop('counter');

  expect(initialSubscriptionValue).toEqual(INITIAL_COUNTER_STATE);
  expect(increaseSubscriptionValue).toEqual(INCREMENTED_COUNTER_STATE);
});

test('Wrapped component re-renders on subscription change ', () => {
  const CounterComponent = Counter;
  const onRender = sinon.spy(CounterComponent.prototype, 'render');

  const store = createSlimReduxStore({
    counter: 0,
    somethingElse: 'test',
  });

  const increaseCounter = store.createChangeTrigger({
    actionType: 'INCREMENT',
    reducer: state => ({...state, counter: store.counter + 1 }),
  });

  const CounterSubscribed = slimReduxReact({
    component: CounterComponent,
    subscriptions: { counter: 'state.counter' }
  });

  const wrapper = mount(
    <div>
      <Provider store={store}>
        <CounterSubscribed/>
      </Provider>
    </div>
  );

  // Save initial render count
  const initialRenderCount = onRender.callCount;

  // Modify subscribed-to state and save call count
  increaseCounter();
  const shouldIncreaseRenderCount = onRender.callCount;

  expect(shouldIncreaseRenderCount).toEqual(initialRenderCount + 1);

  // Cleanup the sinon spy
  onRender.restore();
});

test('Wrapped component does not re-render on change of non-subscribed to data', () => {
  const CounterComponent = Counter;
  const onRender = sinon.spy(CounterComponent.prototype, 'render');

  const store = createSlimReduxStore({
    counter: 0,
    somethingElse: 'test',
  });

  const modifySomethingElse = store.createChangeTrigger({
    actionType: 'MODIFY_SOME_OTHER_PART_OF_THE_STATE',
    reducer: state => ({...state, somethingElse: store.somethingElse + '.' }),
  });

  const CounterSubscribed = slimReduxReact({
    component: CounterComponent,
    subscriptions: { counter: 'state.counter' }
  });

  const wrapper = mount(
    <div>
      <Provider store={store}>
        <CounterSubscribed/>
      </Provider>
    </div>
  );

  // Save initial render count
  const initialRenderCount = onRender.callCount;

  // Modify non-subscribed to state and save call count
  modifySomethingElse();
  const shouldNotChangeRenderCount = onRender.callCount;

  expect(shouldNotChangeRenderCount).toEqual(initialRenderCount);

  // Cleanup the sinon spy
  onRender.restore();
});

test('Change trigger called inside of wrapped component works', () => {
  const CounterComponent = Counter;

  const store = createSlimReduxStore({
    counter: 0,
    somethingElse: 'test',
  });

  const increaseCounter = {
    actionType: 'INCREMENT',
    reducer: state => ({...state, counter: state.counter + 1 }),
  };

  const CounterContainer = slimReduxReact({
    component: CounterComponent,
    subscriptions: { counter: 'state.counter' },
    changeTriggers: { increaseCounter }
  });

  const wrapper = mount(
    <div>
      <Provider store={store}>
        <CounterContainer/>
      </Provider>
    </div>
  );

  // Save initial render count
  wrapper.find('button').simulate('click');
  expect(store.getState().counter).toEqual(1);
});
