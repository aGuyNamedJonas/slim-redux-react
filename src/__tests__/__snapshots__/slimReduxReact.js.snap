// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full 1`] = `
<div>
  <div>
    <h1>
      0
    </h1>
    <button
      onClick={[Function]}
    >
      INCREASE COUNTER
    </button>
  </div>
</div>
`;

exports[`no-change-triggers 1`] = `
<div>
  <div>
    <h1>
      0
    </h1>
    <button
      onClick={[Function]}
    >
      INCREASE COUNTER
    </button>
  </div>
</div>
`;

exports[`no-subscriptions 1`] = `
<div>
  <div>
    <h1 />
    <button
      onClick={[Function]}
    >
      INCREASE COUNTER
    </button>
  </div>
</div>
`;
